File: server/src/sc/server/network/IClientListener.java
package sc.server.network;

import sc.api.plugins.exceptions.RescuableClientException;
import sc.protocol.responses.ProtocolErrorMessage;
import sc.shared.InvalidGameStateException;

public interface IClientListener {

  /**
   * Invoked when a client has disconnected.
   *
   * @param source client, that disconnected
   */
  void onClientDisconnected(Client source);

  /** Invoked when new data is received and ready to be processed. */
  void onRequest(Client source, PacketCallback packet)
          throws RescuableClientException, InvalidGameStateException;

  /**
   * Invoked when a client encountered a problem.
   *
   * @param source client
   * @param packet which caused the problem
   */
  void onError(Client source, ProtocolErrorMessage packet);
}

File: server/src/sc/server/network/IClientListener.kt
package sc.server.network

import sc.networking.clients.XStreamClient.DisconnectCause
import sc.api.plugins.exceptions.RescuableClientException
import sc.shared.InvalidGameStateException

interface IClientListener {
    /** Invoked when this client disconnected.  */
    fun onClientDisconnected(source: Client)
}

interface IClientRequestListener {
    /** Invoked when new data is received and ready to be processed.  */
    @Throws(RescuableClientException::class, InvalidGameStateException::class)
    fun onRequest(source: Client, callback: PacketCallback)
}